asab/web/authn/oauth/cache.py

```diff
@@ -0,0 +1,80 @@
+import time
+import logging
+import collections
+
+import aiohttp
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class OAuthIdentityCache(collections.abc.MutableMapping):
+	"""
+	OAuthIdentityCache serves to store and periodically refresh OAuth identities obtained from the OAuth server
+	specified in the methods list.
+
+	:methods is a list that specifies the identification of OAuth servers such as [asab.web.authn.oauth.GitHubOAuthMethod()]
+	:identity_cache_longevity is an integer that specifies the number of seconds after which the cached identity expires
+	"""
+
+	def __init__(self, app, methods_dict, identity_cache_longevity=60*60):
+		# Prepare empty cache
+		self.IdentityCache = {}
+		self.IdentityCacheLongevity = identity_cache_longevity
+		self.MethodsDict = methods_dict
+
+		# Subscribe to periodically refresh cache
+		app.PubSub.subscribe("Application.tick/10!", self._refresh)
+
+	def __getitem__(self, oauth_server_id_access_token):
+		return self.IdentityCache.get(oauth_server_id_access_token)
+
+	def __setitem__(self, oauth_server_id_access_token, value):
+		if len(value) < 2:
+			raise ValueError("OAuthIdentityCache expends tuple with two items as value.")
+		expiration = time.time() + self.IdentityCacheLongevity
+		self.IdentityCache[oauth_server_id_access_token] = {
+			"OAuthUserInfo": value[0],
+			"Identity": value[1],
+			"Expiration": expiration,
+		}
+
+	def __delitem__(self, oauth_server_id_access_token):
+		if oauth_server_id_access_token in self.IdentityCache:
+			del self.IdentityCache[oauth_server_id_access_token]
+			return True
+		return False
+
+	def __iter__(self):
+		return self.IdentityCache.__iter__()
+
+	def __len__(self):
+		return len(self.IdentityCache)
+
+	async def _refresh(self, message_type):
+		current_time = time.time()
+		for oauth_server_id_access_token, identity in self.IdentityCache.items():
+			expiration = identity.get("Expiration", time.time())
```

Comment:
`time.time()` is expensive operation ... use loop time or call `time.time()` only once.